server/src/tests/evm_contracts/block_env.rs

```diff
@@ -0,0 +1,201 @@
+use {
+    self::evm_contract::BlockEnv::{
+        getBlockNumberCall, getBlockhashCall, getChainIdCall, getMsgSenderCall, getTimestampCall,
+        BlockEnvEvents,
+    },
+    super::*,
+    crate::tests::test_context::TestContext,
+    alloy::{
+        eips::BlockNumberOrTag,
+        primitives::{B256, U256},
+        sol_types::{SolCall, SolEventInterface},
+    },
+    umi_blockchain::receipt::TransactionReceipt,
+};
+
+mod evm_contract {
+    pub const BYTE_CODE_HEX: &str = include_str!("block_env_contract.hex");
+
+    alloy::sol! {
+        #[derive(Debug)]
+        contract BlockEnv {
+            event TheHash (
+                bytes32 indexed hash
+            );
+
+            event TheUint (
+                uint indexed value
+            );
+
+            event TheAddress (
+                address indexed addr
+            );
+
+            function getBlockhash(uint blockNumber) public returns (bytes32);
+            function getBlockNumber() public returns (uint);
+            function getTimestamp() public returns (uint);
+            function getChainId() public returns (uint);
+            function getMsgSender() public returns (address);
+        }
+    }
+}
+
+#[tokio::test]
+async fn test_block_env_evm_contract() -> anyhow::Result<()> {
+    TestContext::run(|mut ctx| async move {
+        // Change the timestamp so that time != block height
+        ctx.timestamp = 1000;
+        let chain_id = ctx.genesis_config.chain_id;
+
+        // 1. Deploy contract
+        let bytecode = alloy::hex::decode(evm_contract::BYTE_CODE_HEX).unwrap();
+        let tx = deploy_evm_contract(chain_id, &bytecode);
+        let receipt = ctx.execute_transaction(tx).await.unwrap();
+        assert!(receipt.inner.inner.is_success());
+        let contract_address = receipt.inner.contract_address.unwrap();
+
+        // 2.a check the `msg.sender` as a transaction
+        let tx = call_contract(
+            chain_id,
+            contract_address,
+            getMsgSenderCall::SELECTOR.to_vec(),
+        );
+        let signer = tx.recover_signer().unwrap();
+        let receipt = ctx.execute_transaction(tx).await.unwrap();
+        let logged_sender = get_logged_address(&receipt);
+        assert_eq!(logged_sender, signer);
+
+        // 2.b check `msg.sender` as a view call
+        let request = view_contract(contract_address, getMsgSenderCall::SELECTOR.to_vec());
+        let sender = request.from.unwrap();
+        let response = ctx
+            .eth_call(request, BlockNumberOrTag::Latest)
+            .await
+            .unwrap();
+        assert_eq!(
+            response,
+            [[0_u8; 12].as_slice(), sender.0.as_slice()].concat()
```

Comment:
nit: This is EVM to MoveVM address conversion, could be done differently but that's OK.